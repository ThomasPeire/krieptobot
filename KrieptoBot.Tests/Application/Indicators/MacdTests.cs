using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FluentAssertions;
using KrieptoBot.Application.Indicators;
using KrieptoBot.DataVisualizer;
using KrieptoBot.DataVisualizer.Extensions;
using KrieptoBot.Domain.Trading.ValueObjects;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Microsoft.FSharp.Core;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Plotly.NET;
using Snapshooter.NUnit;
using Color = System.Drawing.Color;

namespace KrieptoBot.Tests.Application.Indicators
{
    internal class MacdTests
    {
        private IEnumerable<Candle> _candles;

        [SetUp]
        public void Setup()
        {
            InitCandles();
        }

        private void InitCandles()
        {
            var candlesJson = File.ReadAllText(@"./MockData/Bitvavo/candles_btc-eur.json");
            var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
            _candles = deserializedCandles.Select(x =>
                new CandleDto
                {
                    TimeStamp = x.Value<long>(0),
                    Open = x.Value<decimal>(1),
                    High = x.Value<decimal>(2),
                    Low = x.Value<decimal>(3),
                    Close = x.Value<decimal>(4),
                    Volume = x.Value<decimal>(5)
                }.ConvertToKrieptoBotModel()).DistinctBy(x => x.TimeStamp);
        }

        [Test]
        public void Macd_Should_CalculateMacd()
        {
            var datetimeFrom = new DateTime(2021, 11, 01);
            var dateTimeTo = datetimeFrom.AddHours(8);

            var candlesToWorkWith = _candles
                .Where(x => x.TimeStamp >= datetimeFrom && x.TimeStamp <= dateTimeTo)
                .OrderBy(x => x.TimeStamp).ToList();

            var values = new Macd(new ExponentialMovingAverage()).Calculate(candlesToWorkWith);

            values = candlesToWorkWith.ToDictionary(x => x.TimeStamp,
                x => values.TryGetValue(x.TimeStamp, out var value) ? value : 0);

            values.Should().MatchSnapshot();

#if DEBUG
            var candleVisualizer = new CandlesVisualizer();
            var candleChart = candleVisualizer.Visualize(candlesToWorkWith);

            candleChart = candleChart.AddSubChart(values);
            candleChart.WithSize(1920, 1080).WithConfig(Config.init(Responsive: true)).Show();
#endif
        }
    }
}
