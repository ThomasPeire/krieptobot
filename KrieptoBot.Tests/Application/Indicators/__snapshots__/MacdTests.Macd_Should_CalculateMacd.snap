{
  "2021-11-01T00:00:00Z": 0.0,
  "2021-11-01T00:05:00Z": 10.051282051282051282051281,
  "2021-11-01T00:10:00Z": 29.378625173496968368763240,
  "2021-11-01T00:15:00Z": 50.727641862713818298594395,
  "2021-11-01T00:20:00Z": 55.867439646428286838506618,
  "2021-11-01T00:25:00Z": 59.816085321458958351674239,
  "2021-11-01T00:30:00Z": 60.792191359486305453424662,
  "2021-11-01T00:35:00Z": 67.724557157262873063588958,
  "2021-11-01T00:40:00Z": 64.406900079654624062280768,
  "2021-11-01T00:45:00Z": 61.153386413954091379634399,
  "2021-11-01T00:50:00Z": 66.443045092804624561268904,
  "2021-11-01T00:55:00Z": 69.830178871473223212848429,
  "2021-11-01T01:00:00Z": 75.517190477307900524338388,
  "2021-11-01T01:05:00Z": 81.026761502080848269530368,
  "2021-11-01T01:10:00Z": 88.887231547107165183632002,
  "2021-11-01T01:15:00Z": 88.448713986405764872074797,
  "2021-11-01T01:20:00Z": 87.575816023048191808922289,
  "2021-11-01T01:25:00Z": 87.728664444656381667604257,
  "2021-11-01T01:30:00Z": 75.201936975037084472353900,
  "2021-11-01T01:35:00Z": 65.248495943568551519125227,
  "2021-11-01T01:40:00Z": 47.213765836935728817659740,
  "2021-11-01T01:45:00Z": 34.620016997415018965313165,
  "2021-11-01T01:50:00Z": 21.327250618902610745209004,
  "2021-11-01T01:55:00Z": 7.319217124411093071392821,
  "2021-11-01T02:00:00Z": -4.217780253032814998649976,
  "2021-11-01T02:05:00Z": -5.949413181195047076334218,
  "2021-11-01T02:10:00Z": -22.315229142790014691883591,
  "2021-11-01T02:15:00Z": -80.433002569762823909856648,
  "2021-11-01T02:20:00Z": -147.147121322294081001888455,
  "2021-11-01T02:25:00Z": -200.930010700234022244911798,
  "2021-11-01T02:30:00Z": -231.364695988360938919703171,
  "2021-11-01T02:35:00Z": -244.356375960672786013403498,
  "2021-11-01T02:40:00Z": -257.334400144556167229112832,
  "2021-11-01T02:45:00Z": -263.851833335410028099208601,
  "2021-11-01T02:50:00Z": -256.617892577983166637326841,
  "2021-11-01T02:55:00Z": -240.367735089081971285284701,
  "2021-11-01T03:00:00Z": -235.905443779098619128724225,
  "2021-11-01T03:05:00Z": -230.837774590033615369268305,
  "2021-11-01T03:10:00Z": -229.581477936606235040700068,
  "2021-11-01T03:15:00Z": -229.091997904016620993637203,
  "2021-11-01T03:20:00Z": -226.895487930859326557976700,
  "2021-11-01T03:25:00Z": -217.244142647834559389775717,
  "2021-11-01T03:30:00Z": -213.827913236372595831900009,
  "2021-11-01T03:35:00Z": -208.475278646351141809838985,
  "2021-11-01T03:40:00Z": -200.230618740820359977057967,
  "2021-11-01T03:45:00Z": -196.116077575623076661037990,
  "2021-11-01T03:50:00Z": -193.289971790062548972862361,
  "2021-11-01T03:55:00Z": -207.220629971859143530868779,
  "2021-11-01T04:00:00Z": -220.160939031991983785365330,
  "2021-11-01T04:05:00Z": -226.992680443412391576946784,
  "2021-11-01T04:10:00Z": -205.268344735056159600813203,
  "2021-11-01T04:15:00Z": -189.817445308451634153975631,
  "2021-11-01T04:20:00Z": -162.386484842497072802134506,
  "2021-11-01T04:25:00Z": -143.750986074669458377733231,
  "2021-11-01T04:30:00Z": -102.942535631788342137786556,
  "2021-11-01T04:35:00Z": -73.466484308940876797226777,
  "2021-11-01T04:40:00Z": -45.786199263303223019896730,
  "2021-11-01T04:45:00Z": -28.762771386939782333727642,
  "2021-11-01T04:50:00Z": -11.667347593356576299450122,
  "2021-11-01T04:55:00Z": 0.822427146269657217997509,
  "2021-11-01T05:00:00Z": 5.173989382201667441433715,
  "2021-11-01T05:05:00Z": 13.550011198020744768155144,
  "2021-11-01T05:10:00Z": 25.861139273685027633613186,
  "2021-11-01T05:15:00Z": 33.297366235830149221836858,
  "2021-11-01T05:20:00Z": 48.476206548213590361079486,
  "2021-11-01T05:25:00Z": 51.439956575884792503094884,
  "2021-11-01T05:30:00Z": 50.782610226329354950507576,
  "2021-11-01T05:35:00Z": 48.093434479796412652035952,
  "2021-11-01T05:40:00Z": 52.378628017643293613039255,
  "2021-11-01T05:45:00Z": 49.156159957350755407408918,
  "2021-11-01T05:50:00Z": 40.487204525983658461032911,
  "2021-11-01T05:55:00Z": 30.760960760371446796937374,
  "2021-11-01T06:00:00Z": 16.966770223092688976954405,
  "2021-11-01T06:05:00Z": 1.578538090203717021389920,
  "2021-11-01T06:10:00Z": -9.458721137446659491048237,
  "2021-11-01T06:15:00Z": -24.140804906831465482376727,
  "2021-11-01T06:20:00Z": -27.551089285616353760683929,
  "2021-11-01T06:25:00Z": -16.267966342663017183737333,
  "2021-11-01T06:30:00Z": -3.174146756672078074349119,
  "2021-11-01T06:35:00Z": 24.590801884849849468041637,
  "2021-11-01T06:40:00Z": 41.117860770545378460307423,
  "2021-11-01T06:45:00Z": 41.073615045428822788033525,
  "2021-11-01T06:50:00Z": 42.325859961130338169721685,
  "2021-11-01T06:55:00Z": 45.776184496667077136346634,
  "2021-11-01T07:00:00Z": 88.641524978963187983288157,
  "2021-11-01T07:05:00Z": 134.058626211771101746751878,
  "2021-11-01T07:10:00Z": 207.361954799164525900499787,
  "2021-11-01T07:15:00Z": 257.643913609240120027789461,
  "2021-11-01T07:20:00Z": 303.196514801074387563098585,
  "2021-11-01T07:25:00Z": 327.852296192955088480356747,
  "2021-11-01T07:30:00Z": 347.980303879551591039685998,
  "2021-11-01T07:35:00Z": 346.143497321989858389135571,
  "2021-11-01T07:40:00Z": 345.705617570943032427717025,
  "2021-11-01T07:45:00Z": 348.203511013046595763611702,
  "2021-11-01T07:50:00Z": 350.819199481782474778758681,
  "2021-11-01T07:55:00Z": 345.121300709632152357249804,
  "2021-11-01T08:00:00Z": 345.738373743307744165386727
}
